### ISSUE TITLE
rename`rcParams['savefig.directory']` to `rcParams['gui.save.directory']`

### ISSUE THREAD

**Issue by astromancer** [matplotlib/matplotlib#20790]
### Bug summary

Figures saved via the `savefig` method are not saved relative to the `rcParams['savefig.directory']` path. Looking at the source code, it seems that this path is only prepended to the filename if the save request came via the interactive toolbar. Not sure if this is done intentionally, but it seems to me that the `savefig.directory` should always be respected.  

In the example below I use the home directory `~`, but the same behaviour occurs for any other existing directory.

### Code for reproduction

```python
from pathlib import Path
from matplotlib import rc, pyplot as plt


# All figures to go into home
rc('savefig', directory='~')

fig, ax = plt.subplots()
fig.savefig('test.png')

(Path.home() / 'test.png').exists() # False => Not saved in rcParams['savefig.directory']
(Path('.') / 'test.png').exists() # True => Saved in current working directory
```

### Actual outcome

The figure is saved in the local working directory. 

### Expected outcome

If `rcParams['savefig.directory']` has been set, and `savefig` is given a filename (not a full path), the figure should be saved relative to `savefig.directory`.

### Operating system

Ubuntu

### Matplotlib Version

3.3.4

### Matplotlib Backend

Qt5Agg

### Python version

3.8.10

### Jupyter version

NA

### Installation

pip

**Comment by anntzer** [comment]
This behavior is intentional and will almost certainly not change.  Perhaps it could be better documented, though?

**Comment by jklymak** [comment]
I dunno, docs are pretty explicit:
```
# default directory in savefig dialog box,
# leave empty to always use current working directory
```

I'd vote to close this....

**Comment by jklymak** [comment]
I am not sure how the rcParam could be any more explicit, though I agree it was poorly named.  I guess we could deprecate the rcParam and rename it guisavefig....

**Comment by astromancer** [comment]
Well, I have to confess that I didn't come here after reading the `rcParam` docs - I spotted the key and tried to use it intuitively like
```python
rc('savefig', dpi=600, transparent=True, directory='/path/to/folder')
...
fig.savefig('file.png')
```
Which works for the dpi and transparency.

Personally it would be useful to me to be able to set the save directory programmatically. For example, when running a pipeline that produces many figures, its convenient to set the output directory at startup and then subsequently only having to use filenames instead of the full path when saving figures. 


@anntzer
> This behavior is intentional and will almost certainly not change.

Just curious why you see this is as the case? I'm not proposing to change the default behaviour where `rcParam['savefig.directory'] = ''` -  which is what most people use i think. The only change would be when the user has set `rcParam['savefig.directory']` and then provided only a filename to a `savefig` call, the file will be saved in the rcParam directory instead of the current working directory. Calls the come via GUI interactions would be unchanged.


| |||||
| --- | --- | --- | --- | --- |
| **`rcParam['savefig.directory']`**| ''         | ~        | ''         | ~        |
| **`fig.savefig(fname=...)`**      | x.png    | x.png    | /y/x.png | /y/x.png |
| **current outcome**               | ./x.png  | ./x.png  | /y/x.png | /y/x.png |
| **proposed**                      | ./x.png  | **~/x.png**  | /y/x.png | /y/x.png |


What will this require? Inside `savefig`: If `rcParam['savefig.directory']`is non empty and `fname` is path-like and not
>  some backend-dependent object such as `matplotlib.backends.backend_pdf.PdfPages`

then parse it and only if it's not absolute, prepend the rc directory.   I think this will make the API more consistent. Are there people that set `rcParam['savefig.directory']` and still expect `savefig` to save in the local directory? I think not that many, but i could be wrong.

**Comment by QuLogic** [comment]
`savefig.directory` is also used as a runtime location, to store the _last used_ directory for interactive saves, if not empty. Having this suddenly start affecting non-interactive `fig.savefig` would be confusing.

Additionally, `savefig` works like `open` or any other IO function, defaulting to the current working directory. If it started obeying some random configuration, that would quite likely break a bunch of scripts and existing packages.

**Comment by tacaswell** [comment]
I very much agree with @QuLogic that `savefig` departing from the behavior of `open` would be a major API change and is a non-starter.  That said, I also very much agree with @astromancer that the rcparam name is very confusing and suggestive of the proposed behavior.

I think the path forward here is to rename `savefig.directory` to `interactive_savefig.last_used_directory` or move this out of rcparams all together.   I could also see the case for moving to a bool `rc['interactive_save_fig.sticky_last_path']` and then expecting the GUI backends to stash the last used directory someplace at their module level rather than rcparams.

----

This came in via https://github.com/matplotlib/matplotlib/pull/1685 if people want to see the original discussion.

**Comment by timhoffm** [comment]
+1 for renaming. How about `gui.save.directory`?

I think `save[_]fig` in the name is confusing because that will always be associated with the `savefig` method.

I don‘t see a strong case for switching to bool. Giving directory explicitly allows for a special default place to sore images to. I don‘t think we should strip that feature (and I personally use that sometimes). If anything sticky_last_path should be an additional option. But as long as nobody asks for it, I wouldn’t go there.

**Comment by tacaswell** [comment]
I agree we should not drop the feature.  

The strongest argument for switching to bool is that I think this on one of two rcparam that we expect to be changed by the library code under normal operations (the other is 'backend' which we only change once from the secret sentinel to the selected backend so the user can never (without cheating) observe it change) so it is just a bit funny / defies the expectation that "rcparams is a place for users to pass global default parameters to Matplotlib" (by being "a place for Matplotlib to stash some internal state").

**Comment by timhoffm** [comment]
I'd say that "rcparams is the space to store (global) config state".

It's not internal (and also `gui.save.directory` doesn't have to be internal). And yes, users can manipulate that state, with matplotlibrc config files as well as in a local context.

**Comment by tacaswell** [comment]
The distinction I was trying to draw is between the user change the state vs us changing the state.

I'm not going to push this farther and am happy with `gui.save.directory`

**Comment by astromancer** [comment]
Thanks for the insights @QuLogic and @tacaswell. I'll try submit a PR when I have the bandwidth available.

**Comment by Kislovskiy** [comment]
@astromancer I've opened #21370 to address the issue you've pointed out, I hope you are OK with that.
@tacaswell @timhoffm could you take a look?

**Comment by tacaswell** [comment]
This ended up kicking off a long discussion about rcparams on the developer call this week.  We decided that we should close this with no-action as, on balance, we think that the amount of work and time that will be required to make this change in a sufficiently backwards compatible way is not worth the benefit.   

We stand by not making `fig.savefig` or `plt.savefig` look at this rcparam for the reasons above.

------

Because we both validate the rcparams we read from a user's configuration file and the configuration files can not have logic which means users can not do version gating.  Thus, to ensure that users can write a configuration file that work with multiple versions of Matplotlib, we would have to be very careful about how we manage this deprecation. 
 The process would be something like (add an alias in one version, start to warn in a version or two while supporting both, remove the old name) which given our release cadence is best case a 24 month process (1 no warnings, but both exist and work, at least 2 where both exist and work, but old one warns, probably one where both exist, but we ignore the old one, and finally we remove the old one).

This name has been in the code base for ~9 years (36b376134f848503dfdaa74a9bb47217d3b09c6d is where it came in and was first released in 1.3.0) so this is long standing less than ideal name.  For better or worse, our tolerance for bad names goes up the longer we have had the name.  While the name is confusing, there are direct work-arounds (the user keeps track of the directory they want to write to locally in the code or changes cwd for global effect) that are consistent with standard Python behavior and idioms.

Balanced against each other we settled on leaving the name as is.  I am going to close this as no action as we do not want to solicit work on this issue.  

Thank you for reporting this @astromancer , I hope even though we are taking no action you feel we have heard and understood your issue.

----

I made a mistake labeling this "good first issue" over the summer 🐑
